swarm/network/simulation/node_test.go

```diff
@@ -228,7 +229,7 @@ func TestAddNodesAndConnectFull(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	testFull(t, sim, ids)
+	simulations.VerifyFull(t, sim.Net, ids)
```

Comment:
well, there is a bit of redundancy here since the connection and verification of the topology is being done twice (`connect_test.go` and `node_test.go`). we can eliminate those in `node_test.go` then we won't have to export the `VerifyX` functions

Reply:
Would it make sense to move AddNodesAndConnect* functions to the `simulations` package as well, or maybe more or all `simulation` functions? This question of code split suggests that, to me. Package `simulation` was made as an extension to `simulations` and maybe it is good to merge them completely.

Reply:
Well spotted @janos - after browsing through briefly, there doesn't seem to be anything being added in those functions that are dependent of the swarm layer. I agree, all such methods should be moved.

Reply:
@janos we originally did that, and started to move also the `AddNodeX` methods. 

that actually resulted in a huge `Network` type that has a ton of functionality packed into it, and made a mess in the refactor since this actually goes out of band in the question of what needs to be moved and why. 

And so we thought that "hey its probably the Network's responsibility to add nodes" but then another part of the sentence got added - "into the simulation" - there's a whole bunch of stuff in the `AddNodeX` methods that have to be generalised, then propagated into the tests that use the functionality - this is mostly around the instantiation of services. personally i'd be happy to split this into another PR since the delta will be massive. thanks

Reply:
I think that for now, just removing the redundant tests and unexporting the functions would suffice

Reply:
I would not say that the tests are redundant, as `AddNodeAndConnect` function tests do not make any assumptions of implementations that they test. Implementation can change and test ensure that the expected behaviour is preserved. I do not object to remove the tests if you want, but in that case, please, leave a comment of assumptions made and reasons for removing the tests, as someone in the future may want to change function or complain about the fact that they are not tested.

Reply:
I see your @justelad point for not bloating the `simulations` package, that were some of my thought towards having a separate `simulation` package, to keep the api clean. Moving functionalities to `simulations` opens other possibilities for generalizations, but I do not see what you mean with `there's a whole bunch of stuff in the AddNodeX methods that have to be generalised`, they are pretty generalized to me, and cleanly separated from simulations.Network.

Reply:
@janos when I say `AddNodeX` I mean also `AddNode` and `AddNodes`, and then in turn, using the same state of mind - move basically everything in the `node.go` file.
The problem starts in particular with the following lines in `AddNode` function:
```
	if len(conf.Services) == 0 {
		conf.Services = s.serviceNames
	}
```
which is pretty hacky and relies on the fact that `New` method is used and sets the services with some default service entries. This is, to say, not transparent and spaghetti at best.

Needless to say this needs to be refactored and cleaned up but I think that this is out of scope for this PR.

Reply:
I agree that this is a very bad code and spaghetti at best, probably much worse given all the hacks this package contains. Would you create an issue to fix it so that we do not leave it in such state after this PR is closed?

Reply:
jasno 🙏

Reply:
Could it be an idea to have functions without receivers for the add stuff instead?

Reply:
@nolash IMHO we don't need all those methods (i.e. a method that wraps both adding nodes and connecting them). I'd really like to have _less_ functions/methods in this package. If anything we should just have an umbrella function `NewSwarmSimulation(topology simulation.Topology, nodes int, opts AddNodeOption) (*p2p.simulation, error)` that would take care of adding the nodes, connecting them, making sure sanity checks are right then return the simulation for the test.
Then we could get rid of all of the `AddNodesAndConnectX` and all that jazz. less fns/methods more code coverage.